### DISCUSSION TITLE
Should we split `psbt` into a separate crate

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1860] [Ideas]
The `psbt` module depends on code in `rust-bitcoin` but there is no dependency the other way. To reduce the scope of stabalizing the `rust-bitcoin` API would it be useful to spit the `psbt` module out into a separate crate.

## Notes

- The `psbt` module is not nearly as feature complete as the rest of the codebase
- PSBT knowledge is not implied by other Bitcoin knowledge
- @DanGould is hoping to get funding to to work on PSBTv2, he and whomever else gets involved may be able to iterate more quickly if it was separate (this implies though a separate repository)

**Comment by Kixunil** [comment]
We definitely should, you could've kept it an issue. :)

Also all `bip*` modules as their own respective crates.

**Reply by apoelstra** [reply]
BIP32 is a single short module and is pretty integral to Bitcoin wallets, so I think we should keep it in the `bitcoin` crate. It could even go into `primitives`.

Looks like we've renamed BIP143 so it's no longer a `bip*` module, so that's fine :) but 143 is consensus code and definitely needs to stay.

Then BIP 152/158 would probably get more love in a separate crate, yeah.

**Reply by Kixunil** [reply]
I'm not convinced, there's a bunch of applications that don't need BIP32 but do need transactions, addresses, ... But anyway a significant portion of the motivation is to speed up stabilization. BIP32 needs more work, so putting it together with other stuff will slow down the progress.

**Reply by apoelstra** [reply]
Ok, I'm willing to move forward under the assumption that bip32 is going to get separated out. But I would like to assess the amount of work involved to move it back in, and consider doing so, before stabilization.

**Reply by tcharding** [reply]
I'm not convinced that pulling it out just to put it back in is worth the time - given that its me that has to rebase these sorts of PRs eternally.

**Reply by junderw** [reply]
I disagree with removing BIP32.

BIP32 is extremely stable, implemented and crucial to a ton of wallet related APIs in Bitcoin Core, and should be maintained within the core primitives.

**Reply by junderw** [reply]
Similar to how PublicKey and PrivateKey are maintained.

If we have PublicKey and PrivateKey, we should also have bip32, imo.

**Reply by Kixunil** [reply]
I recently came to realize that separating them gives us another benefit: if we eventually decide that we do need to change the API after stabilizing, the damage will be much lower.

I think keys should have their own crate as well. :)

**Reply by junderw** [reply]
If keys are separate that makes sense.

**Comment by apoelstra** [comment]
Agreed, let's not take immediate action to pull it out. Only if we can see a release on the horizon and everything but the split-out is done, and we've made little/no progress on `bip32`, then we should just jettison it for the sake of moving forward.

**Reply by tcharding** [reply]
Cool, got it. Thanks.

**Reply by DanGould** [reply]
I hope a split out would keep the psbt crate in this repo to keep the contribution history and eyeballs for review

**Reply by tcharding** [reply]
Just for the record, as I've done splits before the contribution history gets left in the index of the original crate but I add a comment in the new one saying which commit we "exported" from so one can go back and find it. Eyeballs I can't comment on, we could get less we could get more - who knows :)

**Reply by apoelstra** [reply]
I think `psbt` (and bip32, and probably the other bip modules) should remain in this repository since they're still "core bitcoin concepts" in some sense.

Though I could also see `psbt` growing to `miniscript` levels of complexity and activity, and in that case we may need to separate it. But that won't happen quickly. PSBT2, Musig extensions, Miniscript integration, Payjoin support, etc., ... there's a lot of potential there.

**Comment by sanket1729** [comment]
concept ACK.

**Comment by stevenroose** [comment]
What would be the reason to do this? Like if we do that and also take out primitives, what will be left in `bitcoin`? For some crates, I can see that they have certain reasons to be separate, like minimal deps (hashes), complex and fast-moving APIs (miniscript), way more deps (core-rpc/p2p). bip39 I guess would be an exception and could be part of `bitcoin` IMO, though it has a nice zero-allocations property that would be hard to expose in rust-bitcoin.

**Reply by tcharding** [reply]
The idea was simply about reducing scope of v1.0 stabalisation work. From the thread its not going forward. I marked this as resolved in the title, I've not managed to come up with a better way to indicate that a discussion is resolved.

**Reply by apoelstra** [reply]
BIP39 is never going into rust-bitcoin for the reasons listed in https://secretcodex32.com/faq/index.html

**Reply by Kixunil** [reply]
>  what will be left in `bitcoin`

Reexports of all the separated crates, so `bitcoin` becomes just a facade.

**Reply by tcharding** [reply]
lol, I love how much @apoelstra hates bip39. I'm going to submit a talk to some conf on how good bip39 is just to troll you.

**Comment by tcharding** [comment]
> Reexports of all the separated crates, so bitcoin becomes just a facade.

Is this serious or a joke?

**Reply by stevenroose** [reply]
It's a common model, like crossbeam f.e. But I definitely don't like it. I am for as much as possible in rust-bitcoin unless there is a good reason not to have it.

**Reply by apoelstra** [reply]
We would like to go in the opposite direction, so that downstream users can import the least amount of code that they need from us (and in some cases, avoid pulling in things like rust-secp which can cause pain with wasm), and so that we can stabilize the "easy" parts of the library without needing to stabilize the hard ones.

I don't know that rust-bitcoin will actually make sense as a pure facade but I wouldn't mind too much if it did.

**Comment by tcharding** [comment]
On the roadmap already.
